packages/client/src/symlink.rs

```diff
@@ -165,13 +165,50 @@ impl Symlink {
 		self.resolve_from(tg, None).await
 	}
 
-	#[allow(clippy::unused_async)]
+	#[async_recursion]
 	pub async fn resolve_from(
 		&self,
-		_tg: &dyn Handle,
-		_from: Option<Self>,
+		tg: &dyn Handle,
+		from: Option<Self>,
 	) -> Result<Option<Artifact>> {
-		unimplemented!()
+		let mut from_artifact = if let Some(from) = &from {
+			from.artifact(tg).await?.clone()
+		} else {
+			None
+		};
+		if let Some(artifact::Artifact::Symlink(symlink)) = from_artifact {
+			from_artifact = symlink.resolve(tg).await?;
+		}
```

Comment:
I think this pattern can be cleaned up in rs unlike in ts where you don't have `match` 

```rust
let from_artifact = match from.artifact(tg).await? {
  Some(tg::Artifact::Symlink(symlink)) => Some(symlink.resolve(tg).await?),
  artifact => artifact.clone()
};
let from_path = from.path(tg).await?;

// ... 
let artifact = match self.artifact(tg.await) {
  Some(tg::Artifact::Symlink(symlink)) => Some(symlink.resolve(tg).await?),
  artifact => artifact.clone(),
};
let path = self.path(tg).await?;
```

Then you can do one match 
```rust
match (from_artifact, from_path, artifact, path) {
  (_, _, artifact, None) => Ok(artifact),
  (_, _, artifact, Some(path)) => artifact
    .try_unwrap_directory().wrap_err("Expected a directory")?
    .try_get(path)?
}
```

etc